Userland/Libraries/LibTrash/Trash.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2022, nyabla <hewwo@nyabla.net>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/LexicalPath.h>
+#include <AK/Time.h>
+#include <AK/URL.h>
+#include <LibCore/ConfigFile.h>
+#include <LibCore/File.h>
+#include <LibCore/StandardPaths.h>
+#include <LibCore/System.h>
+#include <LibTrash/Trash.h>
+#include <errno_codes.h>
+
+namespace Trash {
+
+TrashCan::TrashCan()
+{
+    m_trash_directory = String::formatted("{}/.trash", Core::StandardPaths::home_directory());
+    m_info_file_path = String::formatted("{}/trash-info.ini", trash_directory());
+}
+
+ErrorOr<TrashItem> TrashCan::trash(String path, RecursionMode recursive)
+{
+    if (path.is_empty() || !path.starts_with("/"))
+        return Error::from_string_literal("Must be an absolute path");
+    if (!Core::File::exists(path))
+        return Error::from_errno(ENOENT);
+    if (Core::File::is_directory(path) && recursive == RecursionMode::Disallowed)
+        return Error::from_errno(EISDIR);
+    if (m_trash_directory.starts_with(path) || path.starts_with(m_trash_directory))
+        return Error::from_errno(EPERM); //? not sure about this
+
+    i64 timestamp = Time::now_realtime().to_milliseconds();
+    String trash_path = trashed_filename(path, timestamp);
+
+    auto info_file = Core::ConfigFile::open(m_info_file_path, Core::ConfigFile::AllowWriting::Yes);
+    info_file->write_entry(URL::percent_encode(path), String::number(timestamp), "");
```

Comment:
why is percent encoding necessary here? ConfigFile should take care of encoding a key properly so that it can be read again itself.